File: 483.smallest-good-base.go
package leetcode

//
// @lc app=leetcode id=483 lang=golang
//
// [483] Smallest Good Base
//
// Notes
// A good base k represents n as 1 + k + ... + k^p. Try the largest possible
// power p first, because more digits imply a smaller base. For each p, binary
// search the base and evaluate the geometric sum with early stopping. Time:
// O(log^2 n). Space: O(1).
//
// @lc code=start

import "strconv"

func SmallestGoodBase483(n string) string {
	number, _ := strconv.ParseInt(n, 10, 64)
	maxPower := bitLength483(number) - 1

	for power := maxPower; power > 1; power-- {
		left, right := int64(2), number-1
		for left <= right {
			middle := left + (right-left)/2
			currentSum := geometricSum483(middle, power, number)
			if currentSum == number {
				return strconv.FormatInt(middle, 10)
			}
			if currentSum < number {
				left = middle + 1
			} else {
				right = middle - 1
			}
		}
	}

	return strconv.FormatInt(number-1, 10)
}

func geometricSum483(base int64, power int, limit int64) int64 {
	total := int64(1)
	term := int64(1)
	for i := 0; i < power; i++ {
		if term > limit/base {
			return limit + 1
		}
		term *= base
		total += term
		if total > limit {
			break
		}
	}
	return total
}

func bitLength483(value int64) int {
	length := 0
	for value > 0 {
		length++
		value >>= 1
	}
	return length
}

// @lc code=end

File: data structure and algorithms/Math/483.smallest-good-base.java
/*
 * @lc app=leetcode id=483 lang=java
 *
 * [483] Smallest Good Base
 *
 * A good base k represents n as 1 + k + ... + k^power. Try the largest possible
 * power first; the first matching base is smallest. Binary search base and stop
 * geometric summation once it exceeds n.
 *
 * Java note: long is enough for LeetCode's n <= 10^18; overflow is avoided by
 * stopping multiplication when the running sum exceeds the limit.
 *
 * Time: O(log^3 n). Space: O(1).
 */

// @lc code=start
class Solution {
    public String smallestGoodBase(String n) {
        long number = Long.parseLong(n);
        int maxPower = 63 - Long.numberOfLeadingZeros(number) - 1;

        for (int power = maxPower; power > 1; power--) {
            long left = 2;
            long right = number - 1;
            while (left <= right) {
                long mid = left + ((right - left) >>> 1);
                long sum = geometricSum(mid, power, number);
                if (sum == number) {
                    return Long.toString(mid);
                }
                if (sum < number) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return Long.toString(number - 1);
    }

    private long geometricSum(long base, int power, long limit) {
        long total = 1;
        long term = 1;
        for (int i = 0; i < power; i++) {
            if (term > limit / base) {
                return limit + 1;
            }
            term *= base;
            total += term;
            if (total > limit) {
                return total;
            }
        }
        return total;
    }
}
// @lc code=end
